### ISSUE TITLE
What are the WORLD_SIZE and RANK environment variables used for

### ISSUE THREAD

**Issue by davesargrad** [ultralytics/yolov5#3556]
## ❔Question
Hi.
Can you please tell me what the WORLD_SIZE and RANK environment variables are used for:

```
    opt.world_size = int(os.environ['WORLD_SIZE']) if 'WORLD_SIZE' in os.environ else 1
    opt.global_rank = int(os.environ['RANK']) if 'RANK' in os.environ else -1
```

**Comment by glenn-jocher** [comment]
@davesargrad see **Multi-GPU Training Tutorial** below for info on these two arguments.


## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [TorchScript, ONNX, CoreML Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by davesargrad** [comment]
@glenn-jocher 

Hi. 
I did look at that tutorial. It does not seem to provide much detail on those parameters. I'll look again.

**Comment by Paul-Liyb** [comment]
they are used when you need to train on more than 2 GPUs or CPUs. Here may be useful:

**Explanation of environment variables related to torch.ditributed.launch:**

1. **WORLD_SIZE**: In layman's terms, it means how many processes are involved in the training,  WORLD_SIZE = nproc_per_node × nnodes, WORLD_SIZE is unique among different processes;

2. **RANK:** the unique denotation of the process, this value is different in different processes, the above started a total of 8 processes on the two AB machines, then the RANK number of different processes is different

3. **LOCAL_RANK** : LOCAL_RANK is different under the same node, often the GPU is specified according to LOCAL_RANK, but GPU and LOCAL_RANK are not necessarily one-to-one correspondence, because the process is not necessarily limited to the same GPU.

**Comment by glenn-jocher** [comment]
Hello @Paul-Liyb,

The `WORLD_SIZE` and `RANK` environment variables are indeed used in distributed training scenarios. `WORLD_SIZE` refers to the total number of processes participating in the training, which is typically the number of GPUs across all nodes. `RANK` is the unique identifier for each process involved in the training, ranging from 0 to `WORLD_SIZE - 1`.

In the context of YOLOv5, these variables help manage multi-GPU training by coordinating the processes and ensuring they are aware of each other for synchronized operations.

If you're training on a single machine with a single GPU, these variables are not necessary. They come into play when scaling up to multiple GPUs and/or multiple nodes.

I hope this clarifies their purpose! If you have further questions, feel free to ask. 😊
